flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import org.apache.flink.util.FlinkRuntimeException;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The checkpoint failure manager which centralized manage checkpoint failure processing logic.
+ */
+public class CheckpointFailureManager {
+
+	private final static int MAXIMUM_TOLERABLE_FAILURE_NUMBER = Integer.MAX_VALUE;
+
+	private final int tolerableCpFailureNumber;
+	private final FailJobCallback failureCallback;
+	private final AtomicInteger continuousFailureCounter;
+
+	public CheckpointFailureManager(int tolerableCpFailureNumber, FailJobCallback failureCallback) {
+		checkArgument(tolerableCpFailureNumber >= 0
+				&& tolerableCpFailureNumber < MAXIMUM_TOLERABLE_FAILURE_NUMBER,
+			"The tolerable checkpoint failure number is illegal, " +
+				"it must be greater than or equal to 0 and less than " + MAXIMUM_TOLERABLE_FAILURE_NUMBER + ".");
+		this.tolerableCpFailureNumber = tolerableCpFailureNumber;
+		this.continuousFailureCounter = new AtomicInteger(0);
+		this.failureCallback = checkNotNull(failureCallback);
+	}
+
+	/**
+	 * Handle checkpoint exception with a handler callback.
+	 *
+	 * @param exception the checkpoint exception.
+	 * @param checkpointId the failed checkpoint id used to count the continuous failure number based on
+	 *                     checkpoint id sequence. In trigger phase, we may not get the checkpoint id when the failure
+	 *                     happens before the checkpoint id generation. In this case, it will be specified a negative
+	 *                      latest generated checkpoint id as a special flag.
+	 */
+	public void handleCheckpointException(CheckpointException exception, long checkpointId) {
+		CheckpointFailureReason reason = exception.getCheckpointFailureReason();
+		switch (reason) {
+			case PERIODIC_SCHEDULER_SHUTDOWN:
+			case ALREADY_QUEUED:
+			case TOO_MANY_CONCURRENT_CHECKPOINTS:
+			case MINIMUM_TIME_BETWEEN_CHECKPOINTS:
+			case NOT_ALL_REQUIRED_TASKS_RUNNING:
+			case CHECKPOINT_SUBSUMED:
+			case CHECKPOINT_COORDINATOR_SUSPEND:
+			case CHECKPOINT_COORDINATOR_SHUTDOWN:
+			case JOB_FAILURE:
+			case JOB_FAILOVER_REGION:
+				//ignore
+				break;
+
+			case EXCEPTION:
+			case CHECKPOINT_EXPIRED:
+			case CHECKPOINT_DECLINED:
+			case TASK_CHECKPOINT_FAILURE:
+			case TRIGGER_CHECKPOINT_FAILURE:
+			case FINALIZE_CHECKPOINT_FAILURE:
+				continuousFailureCounter.incrementAndGet();
+				break;
+
+			default:
+				throw new FlinkRuntimeException("Unknown checkpoint failure reason : " + reason.name());
+		}
+
+		if (continuousFailureCounter.get() > tolerableCpFailureNumber) {
```

Comment:
The reason I did not use `equal` is here. So if I accept your suggestion, here should change to `continuousFailureCounter.get() > tolerableCpFailureNumber || continuousFailureCounter.get() < 0`. The `continuousFailureCounter.get() < 0` means overflow, right?

Reply:
Exactly. I would even propose that if `tolerableCpFailureNumber == Integer.MAX_VALUE` that we consider this as "unlimited failures allowed" and don't even query the current count, e.g.
`if (tolerableCpFailureNumber != Integer.MAX_VALUE && continuousFailureCounter.get() > tolerableCpFailureNumber)`

Reply:
Accepted! Have updated the code.